File: content/snippets/python/design_patterns/memento.py
class Memento:
    def __init__(self, state: str):
        self.state = state
    
    def get_state(self) -> str:
        return self.state

class Originator:
    def __init__(self):
        self.state = ""
    
    def set_state(self, state: str) -> None:
        self.state = state
    
    def get_state(self) -> str:
        return self.state
    
    def save_state(self) -> Memento:
        return Memento(self.state)
    
    def restore_state(self, memento: Memento) -> None:
        self.state = memento.get_state()

class Caretaker:
    def __init__(self):
        self.history: list[Memento] = []
    
    def save(self, memento: Memento) -> None:
        self.history.append(memento)
    
    def restore(self, index: int) -> Memento:
        if 0 <= index < len(self.history):
            return self.history[index]
        return None

originator = Originator()
caretaker = Caretaker()

originator.set_state("State 1")
caretaker.save(originator.save_state())

originator.set_state("State 2")
caretaker.save(originator.save_state())

originator.set_state("State 3")
print(f"Current: {originator.get_state()}")

originator.restore_state(caretaker.restore(0))
print(f"Restored: {originator.get_state()}")

File: content/snippets/go/design_patterns/memento.go
package main

import "fmt"

type Memento struct {
	state string
}

func (m *Memento) GetState() string {
	return m.state
}

type Originator struct {
	state string
}

func (o *Originator) SetState(s string) {
	o.state = s
}

func (o *Originator) GetState() string {
	return o.state
}

func (o *Originator) SaveState() *Memento {
	return &Memento{o.state}
}

func (o *Originator) RestoreState(m *Memento) {
	o.state = m.GetState()
}

type Caretaker struct {
	history []*Memento
}

func (c *Caretaker) Save(m *Memento) {
	c.history = append(c.history, m)
}

func (c *Caretaker) Restore(index int) *Memento {
	if index >= 0 && index < len(c.history) {
		return c.history[index]
	}
	return nil
}

func main() {
	origin := &Originator{}
	caretaker := &Caretaker{}

	origin.SetState("State 1")
	caretaker.Save(origin.SaveState())

	origin.SetState("State 2")
	caretaker.Save(origin.SaveState())

	origin.SetState("State 3")
	fmt.Printf("Current: %s\n", origin.GetState())

	origin.RestoreState(caretaker.Restore(0))
	fmt.Printf("Restored: %s\n", origin.GetState())
}
